protocols/kad/src/handler.rs

```diff
@@ -468,16 +476,88 @@ pub struct KademliaRequestId {
 struct UniqueConnecId(u64);
 
 impl KademliaHandler {
-    /// Create a [`KademliaHandler`] using the given configuration.
-    pub fn new(
-        config: KademliaHandlerConfig,
+    /// Create a [`KademliaHandler`] for a new inbound connection.
+    pub fn new_inbound(
+        protocol_config: KademliaProtocolConfig,
+        idle_timeout: Duration,
+        local_addr: Multiaddr,
+        remote_addr: Multiaddr,
+        remote_peer_id: PeerId,
+        external_addresses: Vec<&Multiaddr>,
+    ) -> Self {
+        let mode = match external_addresses.as_slice() {
+            [] => {
+                log::debug!("Operating in client-mode on new inbound connection from peer {remote_peer_id} because we don't have any confirmed external addresses");
+
+                Mode::Client
+            }
+            [confirmed_external_addr] => {
+                // This may not be true if we are listening on multiple interfaces but it is a good enough approximation for now.
+                log::debug!("Operating in server-mode on new inbound connection from peer {remote_peer_id} under the assumption that {confirmed_external_addr} routes to {local_addr}");
+
+                Mode::Server
+            }
+            confirmed_external_addresses => {
+                let confirmed_external_addresses = confirmed_external_addresses
+                    .iter()
+                    .map(|addr| addr.to_string())
+                    .collect::<Vec<_>>()
+                    .join(", ");
+
+                log::debug!("Operating in server-mode on new inbound connection from peer {remote_peer_id} under the assumption that one of [{confirmed_external_addresses}] routes to {local_addr}");
+
+                Mode::Server
+            }
+        };
+
+        Self::new(
+            protocol_config,
+            mode,
+            idle_timeout,
+            ConnectedPoint::Listener {
+                local_addr,
+                send_back_addr: remote_addr,
+            },
+            remote_peer_id,
+        )
+    }
+
+    /// Create a [`KademliaHandler`] for a new outbound connection.
+    pub fn new_outbound(
+        protocol_config: KademliaProtocolConfig,
+        idle_timeout: Duration,
+        addr: Multiaddr,
+        role_override: Endpoint,
+        peer: PeerId,
+    ) -> Self {
+        // Outbound connections are always in client-mode because we may be using an ephemeral port for dialing and thus the observed address may not be reachable.
+        log::debug!("Operating in client-mode on new outbound connection to peer {peer} at {addr}");
```

Comment:
I am not sure this is a good idea. Does the go-libp2p Kademlia implementation see the `Mode` as a per-connection property or as a property across all connections? For the sake of consistency and reducing complexity, I suggest treating the `Mode` as a property across all connections. In other words, either the node with all its connections is in `Mode::Server` or `Mode::Client`. Whether the remote node of an outbound connection adds us to their routing table is up to them.

One example would be, what if the local node advertises `Mode::Client` on this connection, but advertises the Kademlia protocol via `libp2p-identify`. How do other implementations cope with this seemingly inconsistent information?

We can still role out this optimization in the future. Though for now, I find treating the `Mode` as a global property much easier to reason about.

Reply:
Let's discuss this in [this](https://github.com/libp2p/rust-libp2p/pull/3877#discussion_r1202347809) threa.d